llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/MaterializeEncoding.cpp

```diff
@@ -256,22 +260,39 @@ lowerOpWithEncoding(RewriterBase &rewriter, linalg::MatmulOp matmulOp,
           mlir::iree_compiler::IREE::LinalgExt::EncodingRole::RESULT) {
     return failure();
   }
-  Operation *mmt4DOp = rewriter.create<linalg::Mmt4DOp>(
-      matmulOp.getLoc(), convertedOutputOperands[0].getType(),
-      convertedInputOperands, convertedOutputOperands);
-  return mmt4DOp;
+
+  FailureOr<MaterializeEncodingInfo> materializeEncodingInfo =
+      materializeEncodingFn(getOriginalTypeWithEncoding(
+          matmulOp.getResultTypes()[0].cast<RankedTensorType>()));
+  Operation *result;
+  if (failed(materializeEncodingInfo)) {
+    SmallVector<Value> operands;
+    operands.append(convertedInputOperands.begin(),
+                    convertedInputOperands.end());
+    operands.append(convertedOutputOperands.begin(),
+                    convertedOutputOperands.end());
+    result = mlir::clone(
+        rewriter, matmulOp,
+        {getTypeWithDroppingEncoding(
+            convertedOutputOperands[0].getType().cast<RankedTensorType>())},
+        operands);
+  } else {
+    result = rewriter.create<linalg::Mmt4DOp>(
+        matmulOp.getLoc(), convertedOutputOperands[0].getType(),
+        convertedInputOperands, convertedOutputOperands);
+  }
+  return result;
```

Comment:
This fallback code looks mostly the same between different op-specific overloads of `lowerOpWithEncoding`, at least between `matmul` and `batch_matmul`, in this file. Can this code be shared in some helper function?